### DISCUSSION TITLE
SQLModel 0.0.14 throws mypy validation errors

### DISCUSSION THREAD

**Discussion by lebaudantoine** [fastapi/sqlmodel#732] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional

from sqlmodel import Field, SQLModel, create_engine

class Hero(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str
    secret_name: str
    age: Optional[int] = None
```

### Description

Recent updates on SQLModel broke mypy linter.
I am using `pydantic==1.10.13` and `mypy==1.7.1`

Linting, the following snippet of code (from the documentation) triggers two mypy errors:

```
...: error: Unexpected keyword argument "table" for "__init_subclass__" of "object"  [call-arg]
...: error: Metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases  [misc]
```

Am  I doing something wrong? 

### Operating System

macOS

### SQLModel Version

0.0.14

### Python Version

3.12

**Comment by kofan** [comment]
Thanks for reporting! I have the same issue with the 0.0.14 version of SQLModel. I had to roll back to the 0.0.12 version... 😔

**Reply by mohit2152sharma** [reply]
faced the same issue, downgrading works, thanks for reporting.

**Comment by hflaime** [comment]
I faced those issues, too. Upgrading to pydantic 2 resolved them for me.

**Comment by kmcentush** [comment]
Does anyone know the right way to fix this when using `mypy` in `pre-commit` hooks? I tried fiddling with `additional_dependencies` but couldn't get anything to work.

Or I guess we wait for a future version where it's fixed in `SQLModel` itself.

**Reply by JonLev** [reply]
Here is what I added for my pre-commit config to make it work (python 3.12.2):
```
-   repo: https://github.com/pre-commit/mirrors-mypy
    rev: v1.8.0
    hooks:
    -   id: mypy
        name: Validate types with MyPy
        types: [python]
        pass_filenames: false
        additional_dependencies:
        -   "pydantic==v2.6.2"
        -   "sqlmodel==0.0.16"
        args:
        -   --config-file=pyproject.toml
```

You need to add missing SQLModel types also

**Reply by kmcentush** [reply]
I'll give this a try later today - are only the `additional_dependencies` needed, and `args` are related to your personal setup? ex: What is `baobab_backend` in the args?

**Reply by JonLev** [reply]
Sorry for the extra arg 😣
The main point is the additional dependencies here, as it will include the type in the pre-commit envs

**Reply by kmcentush** [reply]
Can confirm that adding the two additional dependencies solved my problem. Thanks!

**Reply by AndrewBasem1** [reply]
@JonLev Thank you!

**Reply by menzenski** [reply]
This fixed it for me as well on sqlmodel==0.0.19. Thank you!

**Reply by jbethune** [reply]
> This fixed it for me as well on sqlmodel==0.0.19. Thank you!

Can you also provide the version numbers for your other packages? Thanks!

**Comment by atlekberg** [comment]
I can confirm that this is also an issue with Pyright.

**Comment by Ben-Epstein** [comment]
I'm seeing the same with `sqlmodel==0.0.18`, `mypy==1.10.0`, `pydantic==2.7.1`

**Comment by muhammadyaseen** [comment]
Same issue with the following setup:

```
name         : pydantic                                
version      : 2.7.4

name         : sqlmodel                                                                                   
version      : 0.0.19 

name         : mypy
version       : 1.10.0
```
and adding `additional_dependencies` fixed it.

**Comment by fccoelho** [comment]
facing the same issue, and adding dependencies didn't help.
i am using pydantic 2.8.2, sqlmodel 0.0.21 and mypy 1.11.1

**Reply by juliojm13** [reply]
Using the same versions as you mention and I am facing this issue, any help?
@tiangolo

**Reply by studnikoff** [reply]
the same, please take attention to the issue
@tiangolo

**Comment by ideepu** [comment]
I faced the same issue with the following versions, and adding `additional_dependencies` fixed it for me. I am using mypy with pre-commit.

```
pydantic==2.9.0
sqlmodel==0.0.22
mypy==1.11.2
```

**Reply by Nikola-Milovic** [reply]
What and how did you add the additional dependencies?

**Reply by ideepu** [reply]
I use mypy with pre-commit. This is my config.

```
  - repo: https://github.com/pre-commit/mirrors-mypy
    rev: v1.11.2
    hooks:
      - id: mypy
        additional_dependencies: ["sqlmodel==0.0.22"]
```

**Reply by cguinard-fbx** [reply]
It fixed it for me, I did not even specify the version: `additional_dependencies: ["sqlmodel==0.0.22"]`

**Comment by N-Demir** [comment]
I'm having the same issue
```
Name: mypy
Version: 1.11.1
Name: pydantic
Version: 2.8.2
Name: sqlmodel
Version: 0.0.22
```

**Comment by Cuick** [comment]
Adding an \_\_init\_\_.py file to the relevant directory can resolve this issue.

**Comment by scottgigante-hubflow** [comment]
Issue appears to be with mypy: https://github.com/python/mypy/issues/11036. Adding `sqlmodel` to additional_dependencies does fix it, but it creates many other mypy errors with interactions with sqlalchemy.

**Comment by YuriiMotov** [comment]
Seems to work fine with SQLModel 0.0.24, Pydantic 2.11.7, mypy 1.17.1.
Still shows error with Pydantic V1, but I hope nobody starts new projects with Pydantic V1 nowadays

**Comment by jbw-vtl** [comment]
Also facing this issue, any solution for this yet?

Using python 3.12 and the latest version of all the libraries

**Reply by YuriiMotov** [reply]
Are you using Pydantic V2?
I only managed to reproduce this with Pydantic V2.
Could you share the exact versions of SQLModel and Pydantic libraries. And also, the errors you see

**Reply by jbw-vtl** [reply]
Investigating this further, the issue seems to be linked to running within a pre-commit environment (non local) which doesn't have all the dependencies installed.

I am adding sqlalchemy[mypy] and sqlmodel in the additional dependencies, however the project itself (which is seutp as a local package) is not being imported.
Reviewing this further, not getting the same issue when running directly mypy `locally` in the main env

Will revert

**Comment by thmsklngr-itsv** [comment]
Hi there,
facing the same problem here, Python 3.11.14, Pydantic 2.12.5 and SQLModel 0.0.31. Mypy is at version 1.19.1.
